table/block_based/block_based_table_builder.cc

```diff
@@ -1231,92 +1231,97 @@ void BlockBasedTableBuilder::WriteRawBlock(const Slice& block_contents,
-    io_s = r->file->Append(Slice(trailer.data(), trailer.size()));
-    if (io_s.ok()) {
-      assert(s.ok());
-      bool warm_cache;
-      switch (r->table_options.prepopulate_block_cache) {
-        case BlockBasedTableOptions::PrepopulateBlockCache::kFlushOnly:
-          warm_cache = (r->reason == TableFileCreationReason::kFlush);
-          break;
-        case BlockBasedTableOptions::PrepopulateBlockCache::kDisable:
-          warm_cache = false;
-          break;
-        default:
-          // missing case
-          assert(false);
-          warm_cache = false;
-      }
-      if (warm_cache) {
-        if (type == kNoCompression) {
-          s = InsertBlockInCacheHelper(block_contents, handle, block_type,
-                                       is_top_level_filter_block);
-        } else if (raw_block_contents != nullptr) {
-          s = InsertBlockInCacheHelper(*raw_block_contents, handle, block_type,
-                                       is_top_level_filter_block);
-        }
-        if (!s.ok()) {
-          r->SetStatus(s);
-        }
+
+  {
+    IOStatus io_s = r->file->Append(block_contents);
+    if (!io_s.ok()) {
+      r->SetIOStatus(io_s);
+      r->SetStatus(io_s);
+      return;
+    }
+  }
+
+  std::array<char, kBlockTrailerSize> trailer;
+  trailer[0] = type;
+  uint32_t checksum = ComputeBuiltinChecksumWithLastByte(
+      r->table_options.checksum, block_contents.data(), block_contents.size(),
+      /*last_byte*/ type);
+  EncodeFixed32(trailer.data() + 1, checksum);
+  TEST_SYNC_POINT_CALLBACK(
+      "BlockBasedTableBuilder::WriteRawBlock:TamperWithChecksum",
+      trailer.data());
+  {
+    IOStatus io_s = r->file->Append(Slice(trailer.data(), trailer.size()));
+    if (!io_s.ok()) {
+      r->SetIOStatus(io_s);
+      r->SetStatus(io_s);
+      return;
+    }
+  }
+  {
+    Status s = Status::OK();
+    bool warm_cache;
+    switch (r->table_options.prepopulate_block_cache) {
+      case BlockBasedTableOptions::PrepopulateBlockCache::kFlushOnly:
+        warm_cache = (r->reason == TableFileCreationReason::kFlush);
+        break;
+      case BlockBasedTableOptions::PrepopulateBlockCache::kDisable:
+        warm_cache = false;
+        break;
+      default:
+        // missing case
+        assert(false);
+        warm_cache = false;
+    }
+    if (warm_cache) {
+      if (type == kNoCompression) {
+        s = InsertBlockInCacheHelper(block_contents, handle, block_type,
+                                     is_top_level_filter_block);
+      } else if (raw_block_contents != nullptr) {
+        s = InsertBlockInCacheHelper(*raw_block_contents, handle, block_type,
+                                     is_top_level_filter_block);
       }
-      // TODO:: Should InsertBlockInCompressedCache take into account error from
-      // InsertBlockInCache or ignore and overwrite it.
-      s = InsertBlockInCompressedCache(block_contents, type, handle);
       if (!s.ok()) {
         r->SetStatus(s);
```

Comment:
Note:
This is one of the two places where we don't do early return since we move on to `s = InsertBlockInCompressedCache(block_contents, type, handle);` if `!s.ok()` and someone left a todo about it.

Reply:
Can we resolve the TODO? I think there's no point caching a block for a TableBuilder where non-OK SetStatus() has already happened (i.e., table building already failed).

Reply:
Yep - and let me take a look into the blame on why TODO is left.